docs/spec.md

```diff
@@ -704,195 +708,375 @@ expressions must not be preceded by whitespace before the `[` character,
+## Horizontal whitespace
+
+_Horizontal whitespace_ is a sequence of tabs or spaces that is used to
+separate elements in lists. Lists include the argument list to a
+function call, the element list of an array literal, and the value in
+the key-value pairs of a map literal. However, horizontal whitespace
+is not allowed _within_ this list elements.
+
+Horizontal whitespace is not allowed around the dot expression `.` or
+before the opening brace `[` in an index expression or slice
+expression. However, it is allowed _within_ the grouping expression,
+index expression, and slice expression, even if the expression is an
+element of a list such as an argument to a function call.
+
+More formally, horizontal whitespace `WS` between tokens or terminals as
+defined in the grammar is ignored and can be used freely with the
+addition of the following rules:
+
+1. `WS` is not allowed around dot `.` in dot expressions.
+2. `WS` is not allowed around before the `[` in index or slice expressions.
+3. `WS` is not allowed following the unary operators `-` and `!`.
+4. `WS` is not allowed within arguments to a function call
+5. `WS` is not allowed within elements of an array literal
+6. `WS` is not allowed within the values of a amp literal's key-value pairs.
+7. `WS` is allowed within any grouping expression `(` … `)`.
+8. `WS` is allowed within an index expression `[` … `]`.
+9. `WS` is allowed within a slice expression `[` … `:` … `]`.
+
+Here are some examples of incorrect uses of horizontal whitespace, along
+with their correct uses.
+
+Invalid:
+
+```
+print - 5
+len "a" + "b"
+
+arr := [1 + 1]
+arr [0] = 3 + 2
+print 2 + arr [0]
+
+map := {address: "10 Downing" + "Street"}
+map.  address = "221B Baker Street"
+
+print len map
+```
+
+Valid:
+
+```evy
+print -5
+len "a"+"b"
+
+arr := [1+1]
+arr[0] = 3 + 2
+print 2+arr[0]
+
+map := {address:"10 Downing"+"Street"}
+map.address = "221B Baker Street"
+
+print (len map)
+```
 
 ## Functions
 
-A function declaration binds an identifier, the function name, to a
-function. As part of the function declaration, the function signature
-defines the number, order and types of input parameters as well as the
-result or return type of the function. If the return type is left out
+_Functions_ are blocks of code that are used to perform a specific task.
+They are often used to encapsulate code that is used repeatedly, so
+that it can be called from different parts of a program.
+
+A _function definition_ binds an identifier, the function name, to a
+function. As part of the function definition, the _function signature_
+declares the number, order and types of input parameters as well as the
+result or return type of the function. If the return type is left out,
 the function does not return a value.
 
-    func is_valid:bool text:string cap:num
-        return (len text) <= cap
-    end
+For example, the following code defines a function called `is_valid`
+that takes two parameters, `s` and `cap`, and returns a boolean
```

Comment:
I would change "cap" to something like "max" or "maxLength". Cap (short for capacity I guess) doesn't make a lot of sense for Evy, but a validation function checking a string for a maximum size does make sense.

Reply:
yeah, still haven't made up my mind on snakes vs camels.... max is already taken as a builtin name. ok. i think i've just decided on CamelCase with a heavy hard because I think snake_case is more readable but arguably less writable and less Go-y.